docs/USAGE-GUIDE.md

```diff
@@ -1585,6 +1585,102 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
+int s2n_offered_psk_list_choose_psk(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn, void *context,
+                                          struct s2n_offered_psk_list *psk_list);
+int s2n_config_set_psk_selection_callback(struct s2n_config *config, s2n_psk_selection_callback cb, void *context);
+```
+
+### Configuring External Pre-Shared Keys
+
+Use the following APIs to configure and use external pre-shared keys in the > =TLS1.3 connections. An external PSK is an out-of-band provisioned pre-shared key (PSK) established externally using a secure mutually agreed upon mechanism.
+
+To create an external PSK a caller must call the `s2n_external_psk_new` API.**s2n_external_psk_new** creates a new s2n external pre-shared key (PSK) object with `S2N_PSK_HMAC_SHA256` as the default pre-shared key hash algorithm. Note that the obfuscated ticket age will always be zero for external PSKs. Use `s2n_psk_free` to free the memory allocated to the s2n external pre-shared key object created by this API.
+
+**s2n_psk_free** frees the memory associated with the `s2n_psk` object.
+
+After creation of the external PSK object use `s2n_psk_set_identity`, `s2n_psk_set_secret` to set the PSK Identity and PSK secret. To override the default PSK Hash Algorithm  use `s2n_psk_set_hmac`.
+
+**s2n_psk_set_identity** sets the pre-shared key (PSK) identity for a given s2n_psk object. The pre-shared key identity is a unique identifier for the pre-shared secret. The Pre-shared key identity is transmitted over the network unencrypted and is a non-secret value. Be sure to not include any confidential information as a part of the pre-shared key identity. Note that the identity value is copied into s2n-tls memory and the user is responsbile for freeing the memory  associated with the `identity` input. Be sure to call `free(identity)` after this API call. 
+
+**s2n_psk_set_secret** sets the out-of-band/externally provisioned pre-shared secret (PSK) for a given s2n_psk object. Note that the secret value is copied into s2n-tls memory and the user is responsbile for freeing the memory associated with the `secret` input. Be sure to call `free(secret)` after this API call.
+
+**s2n_psk_set_hmac** sets the PSK HMAC algorithm for a given PSK object.
+The supported pre-shared key HMAC algorithms are listed in the `s2n_psk_hmac` enum. Note that in versions >= TLS1.3 the cipher suite selected has the same Hash alogrithm as the pre-shared key. Use this API to override the default PSK Hash Algorithm value `S2N_PSK_HMAC_SHA256`. 
+
+Once the external PSK object is created and updated to include the PSK identity and secret, the caller can add the external PSK to the connection using `s2n_connection_append_psk` API. This API sets up the list of PSKs on the connection that will later be used by the pre-shared key extension to select a pre-shared key for the connection.
```

Comment:
```suggestion
After adding an identity and secret to the PSK object, append the PSK to the connection using 's2n_connection_append_psk'.
```
I don't really think you need that last sentence, the user doesn't need to know about the psk extension.

Reply:
I think we should mention that the reason we append: `After adding an identity and secret to the PSK object, append the PSK to the connection using 's2n_connection_append_psk'. This API sets up the list of PSKs on the connection that will later be used by the PSK extension to select a PSK for the connection.`